packages/teuchos/CMakeLists.txt

```diff
@@ -349,6 +349,13 @@ ENDIF() # ${PACKAGE_NAME}_ENABLE_COMPLEX
 # C) Set up package-specific options
 #
 
+TRIBITS_ADD_OPTION_AND_DEFINE(
+  ${PACKAGE_NAME}_MODIFY_DEFAULTS_DURING_VALIDATION
+  HAVE_TEUCHOS_MODIFY_DEFAULTS_DURING_VALIDATION
+  "Modify default parameters during validation."
```

Comment:
CC: @dsjense 

@rppawlo, agreed, validating the defaults does break backward compatibility but it seems like we should strive to make that the default at the first chance we can get.

Just adding a comment does not seem like it is enough as it will be missed once Trilinos looks to put out release 16.0.

I am wondering if there might be a mechanism where we could define a standard CMake variable to transition to the next major release and break backward compatibility?  That would also allow customers to test their configurations and usage of Trilinos for the next major release against changes in behavior.  For example, there could be a cache variable named something like `Trilinos_ENABLE_NEW_BEHAVIOR_FOR_NEXT_MAJOR_RELEASE` (with default `OFF`) and then we could define the above code like:

```
TRIBITS_ADD_OPTION_AND_DEFINE(
  ${PACKAGE_NAME}_MODIFY_DEFAULTS_DURING_VALIDATION
  HAVE_TEUCHOS_MODIFY_DEFAULTS_DURING_VALIDATION
  "Modify default parameters during validation."
  ${Trilinos_ENABLE_NEW_BEHAVIOR_FOR_NEXT_MAJOR_RELEASE}
  )
```

That way:

* We can have one or more PR or Nightly build that sets `Trilinos_ENABLE_NEW_BEHAVIOR_FOR_NEXT_MAJOR_RELEASE=ON` and make sure all of Trilinos works with the new behavior so the next major release can be put out quickly.
* Users could set `Trilinos_ENABLE_NEW_BEHAVIOR_FOR_NEXT_MAJOR_RELEASE=ON` so as to get the new behavior early and/or to ensure they are ready for the new behavior in the next major release.
* After the last release in the current major release set is put out, we could search for `Trilinos_ENABLE_NEW_BEHAVIOR_FOR_NEXT_MAJOR_RELEASE` and update the code to only use the new behavior.

I think it might be good to add this variable to the TriBITS support for managing backward compatibility defined at:

* https://tribitspub.github.io/TriBITS/users_guide/index.html#regulated-backward-compatibility-and-deprecated-code

Currently, there is nothing defined there to handle changes in behavior of existing code.  It only addresses deprecating and dropping old code and making users rely on new code.

Again, I just fear that a comment will get ignored and the new behavior will never get enabled by default for new users.

Reply:
After this commit, the default behavior is the new/correct behavior. I talked with Trilinos leadership and since this is really a bug, they are fine with us making this backward incompatible change now. The chances of this this being an issue are low - not many teams use the StringToInt validator. In fact most use of it was incorrect and should really have been using the string validator. The flag was added because two application codes do use the old/wrong behavior (and propagated it all over their code). Those apps can set the new flag in this PR to OFF to get the old behavior with the understanding that the flag will eventually go away.  I can bring up the new flags/capability you mentioned at the next operational leaders meeting.